astropy/table/table.py

```diff
@@ -618,6 +623,28 @@ def __init__(self, data=None, masked=False, names=None, dtype=None,
         if self.masked not in (None, True, False):
             raise TypeError("masked property must be None, True or False")
 
+        self._set_column_attribute('unit', units)
+        self._set_column_attribute('description', descriptions)
+
+    def _set_column_attribute(self, attr, values):
+        """Set ``attr`` for columns to ``values``, which can be either a dict (keyed by column
+        name) or a dict of name: value pairs.
+        """
+        if not values:
+            return
+
+        if isinstance(values, dict):
+            for name, value in values.items():
+                if name not in self.columns:
```

Comment:
I'd move this out of the loop and change to check that all names are column names - that way one can never change only some of the columns (since these may not be copies), i.e.,
```
if any(name not in self.columns):
    raise ...
```

Reply:
Well, if the exception gets raised then there is no table returned nothing is partially done.  Doing it outside the loop makes it harder to know which one is bad.

Reply:
Do we always make copies of the columns? (not the data, just the instance) If so, indeed, no harm done.

Reply:
Yes, it always gets a copy of a column.  Note that it can get a reference to a Quantity, but in this case you can't change the unit anyway so I think we're OK.

Reply:
OK, let's stick with what you have, then!